### PR TITLE
Issue #1973 - Implement minimum response data rate

### PR DESCRIPTION
**Pull request by sbordet**

### CODE REVIEW

**Comment by sbordet** [comment]
@gregw: commit 4e90534 has the #1973 implementation, the previous commits are just cleanups.

File: jetty-io/src/main/java/org/eclipse/jetty/io/WriteFlusher.java [hunk 1]
```diff
@@ -104,29 +103,30 @@ protected WriteFlusher(EndPoint endPoint)
...
-            LOG.debug("update {}:{}{}{}", this, previous, updated?"-->":"!->",next);
+            LOG.debug("update {}:{}{}{}", this, previous, updated ? "-->" : "!->", next);
         return updated;
     }
 
     private void fail(PendingState pending)
     {
         State current = _state.get();
-        if (current.getType()==StateType.FAILED)
+        if (current.getType() == StateType.FAILED)
```
Comment (gregw) [inline jetty-io/src/main/java/org/eclipse/jetty/io/WriteFlusher.java]: Please don't do reformat changes with substantial changes!  it makes it impossible to see what has changed!

File: jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Flusher.java [hunk 2]
```diff
@@ -199,6 +201,29 @@ protected Action process() throws Throwable
         return Action.SCHEDULED;
     }
 
+    void onFlushed(long bytes) throws IOException
```
Comment (gregw) [inline jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Flusher.java]: some comments on this method would be good.  It's the main complexity.

File: jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Session.java [hunk 3]
```diff
@@ -1239,7 +1257,8 @@ public void succeeded()
     private class DataEntry extends HTTP2Flusher.Entry
     {
         private int bytes;
-        private int dataRemaining;
+        private int frameBytes;
+        private int dataBytes;
```
Comment (gregw) [inline jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Session.java]: frameBytes and dataBytes are not good names.  Is this in total? a progressive value?  So a better name or a comment

File: jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Session.java [hunk 4]
```diff
@@ -1250,35 +1269,47 @@ private DataEntry(DataFrame frame, IStream stream, Callback callback)
             // of data frames that cannot be completely written due to
             // the flow control window exhausting, since in that case
             // we would have to count the padding only once.
-            dataRemaining = frame.remaining();
+            dataBytes = frame.remaining();
+        }
+
+        @Override
+        public int getFrameBytesLeft()
```
Comment (gregw) [inline jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Session.java]: use `Remaining` rather than `Left`

Comment (gregw) [reply jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Session.java]: Actually is it Remaining?  updateFrameBytes adds updates, so it is how many have been written, not how many remaining to write?

File: jetty-server/src/main/java/org/eclipse/jetty/server/HttpOutput.java [hunk 5]
```diff
@@ -231,6 +229,8 @@ private void write(ByteBuffer content, boolean complete) throws IOException
 
     protected void write(ByteBuffer content, boolean complete, Callback callback)
     {
+        if (_firstByteTimeStamp == -1)
+            _firstByteTimeStamp = System.nanoTime();
```
Comment (gregw) [inline jetty-server/src/main/java/org/eclipse/jetty/server/HttpOutput.java]: Should we check the getHttpConfiguration().getMinResponseDataRate() here and only set the timestamp if we need it. That way onFlushed only need check _firstByteTimeStamp and it knows it has a minResponseDataRate!

File: jetty-server/src/main/java/org/eclipse/jetty/server/HttpOutput.java [hunk 6]
```diff
@@ -908,6 +908,22 @@ public void setBufferSize(int size)
         _commitSize = size;
     }
 
+    public void onFlushed(long bytes) throws IOException
+    {
```
Comment (gregw) [inline jetty-server/src/main/java/org/eclipse/jetty/server/HttpOutput.java]: probably should comment that the bytes flushed here could be after compression and thus less than the bytes written

File: jetty-io/src/main/java/org/eclipse/jetty/io/WriteFlusher.java [hunk 7]
```diff
@@ -550,4 +555,16 @@ public String toStateString()
...
+    @Override
+    public String toString()
+    {
+        State s = _state.get();
+        return String.format("WriteFlusher@%x{%s}->%s", hashCode(), s, s instanceof PendingState ? ((PendingState)s).getCallback() : null);
+    }
+
+    public interface Listener
+    {
+        void onFlushed(long bytes) throws IOException;
```
Comment (gregw) [inline jetty-io/src/main/java/org/eclipse/jetty/io/WriteFlusher.java]: I like the nice simple signature.
But needs comments to say when it is called, under what concurrency and what happens to the exceptions thrown.

**Comment by sbordet** [comment]
@gregw feel free to re-review and eventually merge.

**Pull request merged**